### PR TITLE
#250760 [feature] Mercure Integration

### PR DESCRIPTION
**Pull request by chetan-thapliyal**

### CODE REVIEW

File: src/Mealz/MealBundle/Controller/ParticipantController.php [hunk 1]
```diff
@@ -246,4 +260,38 @@ private function generateResponse(string $route, string $action, Participant $pa
...
+            $combinedDishes = $participant->getCombinedDishes();
+            /** @var Dish $dish */
+            foreach ($combinedDishes as $dish) {
+                $dishTitle .= ' - ' . $dish->getTitleEn();
+            }
+        }
+
+        // If the meal has variations, get its parent and concatenate
+        // the title of the parent meal with the title of the variation.
+        if ($bookedDish->getParent()) {
```
Comment (hacksch) [inline src/Mealz/MealBundle/Controller/ParticipantController.php:279]: check for data type?

File: src/Mealz/MealBundle/Event/MealOfferedEvent.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+declare(strict_types=1);
+
+namespace App\Mealz\MealBundle\Event;
+
+use App\Mealz\MealBundle\Entity\Meal;
+use Symfony\Contracts\EventDispatcher\Event;
+
+class MealOfferedEvent extends Event
+{
+    private Meal $Meal;
```
Comment (hacksch) [inline src/Mealz/MealBundle/Event/MealOfferedEvent.php:12]: why is the first letter uppercase?

File: src/Mealz/MealBundle/Event/Subscriber/MealOfferSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+        $this->publish([
+            'state' => 'gone',
+            'mealId' => $meal->getId(),
+        ]);
+    }
+
+    private function publish(array $data): void
+    {
+        $published = $this->publisher->publish(self::TOPIC_MEAL_OFFER_UPDATES, $data);
```
Comment (hacksch) [inline src/Mealz/MealBundle/Event/Subscriber/MealOfferSubscriber.php:98]: maybe rename it to isPublished

File: src/Mealz/MealBundle/Service/OfferService.php [hunk 4]
```diff
@@ -55,4 +64,20 @@ public static function updateOfferCount(array &$offers, string $key, array $dish
...
+    /**
+     * Gets count of offered meals on $date.
+     */
+    public function getOfferCount(DateTime $date): int
+    {
+        return $this->participantRepo->getOfferCount($date);
+    }
+
+    /**
+     * Gets count of offered meals on $date.
```
Comment (hacksch) [inline src/Mealz/MealBundle/Service/OfferService.php:77]: comment seems wrong

File: src/Resources/js/modules/participation-update-handler.ts [hunk 5]
```diff
@@ -19,192 +22,418 @@ export interface AcceptOfferData {
...
+            } else {
+                $mealContainer.attr('data-available-dishes', update.availableWith.join(','))
+            }
+        }
     }
 
-    public static acceptOffer($checkbox: JQuery, data: AcceptOfferData): void {
-        this.changeToSwapState($checkbox, data.url, data.participantsCount);
+    public static changeToAssignedSlot($checkbox: JQuery, slot: string){
+        let $assignedSlot = $checkbox.closest('.meal').find('[value='+ slot +']');
```
Comment (hacksch) [inline src/Resources/js/modules/participation-update-handler.ts:114]: would write it without extra variable

$checkbox.closest('.meal')
  .find('[value='+ slot +']')
  .prop('selected', 'selected');

File: src/Resources/js/modules/participation-update-handler.ts [hunk 6]
```diff
@@ -19,192 +22,418 @@ export interface AcceptOfferData {
...
+            $tooltip.toggleClass('active');
+        }
+    }
 
-        // update
-        this.updateCheckboxEnabled($checkbox);
-        this.updateCheckBoxWrapper($checkbox);
-        this.toggleTooltip($checkbox);
+    private static updateCheckboxWrapper($checkbox: JQuery) {
+        let $checkboxWrapper = $checkbox.closest('.checkbox-wrapper');
```
Comment (hacksch) [inline src/Resources/js/modules/participation-update-handler.ts:379]: why not write like this?

$checkbox.closest('.checkbox-wrapper')
       .toggleClass('checked', $checkbox.is(':checked'))
       .toggleClass('disabled', $checkbox.is(':disabled'));

File: src/Resources/js/modules/slot-allocation-update-handler.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            let $slotOption = $slotSelector.find(`option[value="${slot}"]`);
+            if (1 !== $slotOption.length) {
+                continue;
+            }
+
+            const slotLimit = $slotOption.data('limit');
+            if (slotLimit > 0) {
+                const slotTitle = $slotOption.data('title');
+                const slotText = `${slotTitle} (${count}/${slotLimit})`;
+                $slotOption.text(slotText);
```
Comment (hacksch) [inline src/Resources/js/modules/slot-allocation-update-handler.ts:26]: can be combined

$slotOption.text(slotText)
                .prop('disabled', slotLimit <= count);

File: src/Resources/js/views/mealGuest.ts [hunk 8]
```diff
@@ -14,14 +16,16 @@ export default class MealGuestView {
...
-            setInterval(this.updateSlots.bind(this), 3000);
         }
 
         if (this.$participationCheckboxes.length > 0) {
             let participationToggleHandler = new ParticipationGuestToggleHandler(this.$participationCheckboxes);
             new ParticipationPreToggleHandler(participationToggleHandler);
-            new ParticipationGuestCountUpdateHandler(this.$participationCheckboxes);
         }
+
+        let messageSubscriber = new MercureSubscriber($('[data-msg-subscribe-url]').data('msgSubscribeUrl'));
```
Comment (hacksch) [inline src/Resources/js/views/mealGuest.ts:26]: new MercureSubscriber($('[data-msg-subscribe-url]')
   .data('msgSubscribeUrl'))
   .subscribe(['participation-updates'], ParticipationUpdateHandler.updateParticipation)
   .subscribe(['slot-allocation-updates'], SlotAllocationUpdateHandler.handleUpdate);

File: src/Resources/js/views/mealIndex.ts [hunk 9]
```diff
@@ -1,39 +1,54 @@
...
 
         if (this.$participationCheckboxes.length > 0) {
             let participationToggleHandler = new ParticipationToggleHandler(this.$participationCheckboxes);
             this.participationPreToggleHandler = new ParticipationPreToggleHandler(participationToggleHandler);
-
-            new ParticipationCountUpdateHandler(this.$participationCheckboxes);
-            new UpdateOffersHandler();
         }
+
+        let messageSubscriber = new MercureSubscriber($('.weeks').data('msgSubscribeUrl'));
```
Comment (hacksch) [inline src/Resources/js/views/mealIndex.ts:29]: can be written without declaration of variable

File: src/Mealz/MealBundle/Controller/ParticipantController.php (hunk 1 above)
Comment (chetan-thapliyal) [reply src/Mealz/MealBundle/Controller/ParticipantController.php]: Fixed

File: src/Mealz/MealBundle/Event/MealOfferedEvent.php (hunk 2 above)
Comment (chetan-thapliyal) [reply src/Mealz/MealBundle/Event/MealOfferedEvent.php:12]: Fixed

File: src/Mealz/MealBundle/Service/OfferService.php (hunk 4 above)
Comment (chetan-thapliyal) [reply src/Mealz/MealBundle/Service/OfferService.php:77]: Fixed

File: src/Resources/js/views/mealGuest.ts (hunk 8 above)
Comment (chetan-thapliyal) [reply src/Resources/js/views/mealGuest.ts:18]: 1. "$('[data-msg-subscribe-url]').data('msgSubscribeUrl')" is the parameter to the constructor.
2. Subscribe is not a jquery method.

File: src/Resources/js/views/mealIndex.ts (hunk 9 above)
Comment (chetan-thapliyal) [reply src/Resources/js/views/mealIndex.ts:30]: Can you give an example?

File: src/Resources/js/modules/slot-allocation-update-handler.ts (hunk 7 above)
Comment (chetan-thapliyal) [reply src/Resources/js/modules/slot-allocation-update-handler.ts:26]: Fixed

File: src/Resources/js/modules/participation-update-handler.ts (hunk 6 above)
Comment (chetan-thapliyal) [reply src/Resources/js/modules/participation-update-handler.ts:389]: Updated

**Pull request merged**